clients/include/testing_csx2dense.hpp

```diff
@@ -0,0 +1,396 @@
+    {
+        CHECK_HIPSPARSE_ERROR(dense2csx(handle,
+                                        M,
+                                        N,
+                                        descr,
+                                        (const T*)d_dense_val,
+                                        LD,
+                                        d_nnzPerRowColumn,
+                                        d_csx_val,
+                                        d_csx_row_col_ptr,
+                                        d_csx_col_row_ind));
+    }
+    else
+    {
+        CHECK_HIPSPARSE_ERROR(dense2csx(handle,
+                                        M,
+                                        N,
+                                        descr,
+                                        (const T*)d_dense_val,
+                                        LD,
+                                        d_nnzPerRowColumn,
+                                        d_csx_val,
+                                        d_csx_col_row_ind,
+                                        d_csx_row_col_ptr));
+    }
+
+    //
+    // Copy on host.
+    //
+    CHECK_HIP_ERROR(hipMemcpy(
+        cpu_csx_val.data(), d_csx_val, sizeof(T) * std::max(nnz, 1), hipMemcpyDeviceToHost));
+
+    CHECK_HIP_ERROR(hipMemcpy(cpu_csx_row_col_ptr.data(),
+                              d_csx_row_col_ptr,
+                              sizeof(int) * (DIMDIR + 1),
+                              hipMemcpyDeviceToHost));
+
+    CHECK_HIP_ERROR(hipMemcpy(cpu_csx_col_row_ind.data(),
+                              d_csx_col_row_ind,
+                              sizeof(int) * std::max(nnz, 1),
+                              hipMemcpyDeviceToHost));
+
+    if(argus.unit_check)
+    {
+        for(int i = 0; i < LD; ++i)
+        {
+            for(int j = 0; j < N; ++j)
+            {
+                h_dense_val[j * LD + i] = T(-2);
+            }
+        }
+        CHECK_HIP_ERROR(
+            hipMemcpy(d_dense_val, h_dense_val.data(), sizeof(T) * LD * N, hipMemcpyHostToDevice));
+        host_csx2dense<DIRA, T>(M,
+                                N,
+                                hipsparseGetMatIndexBase(descr),
+                                (const T*)cpu_csx_val.data(),
+                                cpu_csx_row_col_ptr.data(),
+                                cpu_csx_col_row_ind.data(),
+                                (T*)h_dense_val.data(),
+                                LD);
+
+        if(DIRA == HIPSPARSE_DIRECTION_ROW)
+        {
+            CHECK_HIPSPARSE_ERROR(csx2dense(handle,
+                                            M,
+                                            N,
+                                            descr,
+                                            (const T*)d_csx_val,
+                                            d_csx_row_col_ptr,
+                                            d_csx_col_row_ind,
+                                            (T*)d_dense_val,
```

Comment:
d_dense_val is already of type T* so I dont think you need the cast

Reply:
yes, residual from device_vector usage

Reply:
good point, removed these unnecessary casts.